### ISSUE TITLE
Deadlock issue

### ISSUE THREAD

**Issue by aoeui** [tomerfiliba/rpyc#152]
Currently, when a sync_request is issued, it RPyC alternates between checking if a reply is available and serving new requests. This induces deadlocks in code that I'm trying to utilize with RPyC. Example:

Let R1 be the first request, which triggers a callback that is translated into a sync_request S. Let R2 be a second request. Suppose that R2 _waits_ for S to complete in its code by inspecting some state that is modified by S.

Since RPyC execution alternates, it could potentially serve R2 before S completes. Since S's result cannot propagate due to alternation, this results in a deadlock. S is blocked because R2 is being served. R2 is blocked because it needs the result of S to propagate.

I have patched my local copy to spawn a thread to serve in parallel while looking at sync_request. The solution here lets R2 run in parallel with propagation of S, avoiding the deadlock. I'm not sure if this approach is safe, however.

What my patched code looks like:

```
    seq = self._send_request(handler, args)
    # Keep serving until the synchronous reply has been received.
    reply_received = False
    def serve_until_reply_received():
        while not reply_received:
            self.serve()
    Thread(target=serve_until_reply_received).start()
    while seq not in self._sync_replies:
        time.sleep(0.02)
    reply_received = True
    isexc, obj = self._sync_replies.pop(seq)
    if isexc:
        raise obj
    else:
        return obj
```

**Comment by geromueller** [comment]
hey, should be fixed in 7abaf580fecd0d19ce1383559d1cf41c4688bf18

**Comment by coldfix** [comment]
Closing assuming it is fixed. Please reopen if the issue remains.
